### PR TITLE
Write immutable samplers to descriptor set under and option

### PR DESCRIPTION
**Pull request by s-perron**
This has two places where the immutable samplers could be written.  This
is not the final code.  It is just something to start discussion.  Let
me know what you think.

Discussion for https://github.com/GPUOpen-Drivers/xgl/issues/128.

### CODE REVIEW

File: icd/api/vk_descriptor_set.cpp [hunk 1]
```diff
@@ -87,6 +88,18 @@ void DescriptorSet<numPalDevices>::Reassign(
...
+    for (uint32_t deviceIdx = 0; deviceIdx < numPalDevices; deviceIdx++)
+    {
+        for (uint32_t bindingIndex = 0; bindingIndex < pLayout->Info().count; ++bindingIndex) {
+          const DescriptorSetLayout::BindingInfo& bindingInfo = pLayout->Binding(bindingIndex);
+
+          if (bindingInfo.imm.dwSize == 0) continue;
+
+          uint32_t * pSamplerDesc =(pLayout->Info().imm.pImmutableSamplerData + bindingInfo.imm.dwOffset);
+          uint32_t* pDestAddr = StaticCpuAddress(deviceIdx) + Layout()->GetDstStaOffset(bindingInfo, 0);
+          memcpy(pDestAddr, pSamplerDesc, 4*bindingInfo.imm.dwSize);
```
Comment (s-perron) [inline icd/api/vk_descriptor_set.cpp:100]: Writing the sampler in when assigning the layout to the descriptor set makes the most sense to me.  However, it does require the device to be passed in.

File: icd/api/vk_descriptor_set.cpp [hunk 2]
```diff
@@ -454,12 +467,10 @@ void DescriptorUpdate::WriteDescriptorSets(
...
         case VK_DESCRIPTOR_TYPE_SAMPLER:
-            if (hasImmutableSampler)
-            {
-                VK_ASSERT(!"Immutable samplers cannot be updated");
-            }
-            else
-            {
+            if(params.pImageInfo->sampler == VK_NULL_HANDLE && hasImmutableSampler) {
+              uint32_t * pSamplerDesc =(pDestSet->Layout()->Info().imm.pImmutableSamplerData + destBinding.imm.dwOffset);
+              memcpy(pDestAddr, pSamplerDesc, 4*destBinding.imm.dwSize);
```
Comment (s-perron) [inline icd/api/vk_descriptor_set.cpp:472]: In the cases I was looking at, it seemed like there was a write to the immutable sampler.  I did not trace where this came from.  If it is coming from the application, then it is probably an error on their part.  If not, then there might be something wrong  in the XGL code, but I'm surprised this assert is not being hit more often.

I just want to clarify that that a descriptor write for an immutable sampler is an error.

**Comment by amdvlk-admin** [comment]
### Test summary for commit e54f755b57b9adc399bc729e7d1074fc53f568b8
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev 1ab3bfdf15859e11cd3d609e9886b0393038c980</li>
<li>PAL: dev cd4f8b28b5d923e8d65a1ef1d3dee5137a89d6bd</li>
<li>SPVGEN: dev 8705e87a8b4aaed55a89f79ed423612ca310bfdd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 02a00c5cd4463687bf52d043ed3b7caa14374689</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228848)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32108/57212 (56.1%)</li>
<li>Failed: 0/57212 (0.0%)</li>
<li>Not Supported: 25104/57212 (43.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32108/57212 (56.1%)</li>
<li>Failed: 0/57212 (0.0%)</li>
<li>Not Supported: 25104/57212 (43.9%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32488/57212 (56.8%)</li>
<li>Failed: 0/57212 (0.0%)</li>
<li>Not Supported: 24724/57212 (43.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35419/57212 (61.9%)</li>
<li>Failed: 0/57212 (0.0%)</li>
<li>Not Supported: 21793/57212 (38.1%)</li>
</ul></details>


</ul></details>

File: icd/api/vk_descriptor_set.cpp (hunk 2 above)
Comment (JonCampbell407) [reply icd/api/vk_descriptor_set.cpp:472]: I think it is clear in the spec that an application updating an immutable sampler is an error:

Immutable samplers are permanently bound into the set layout and must not be changed; updating a VK_DESCRIPTOR_TYPE_SAMPLER descriptor with immutable samplers is not allowed

File: icd/api/vk_descriptor_set.cpp (hunk 1 above)
Comment (JonCampbell407) [reply icd/api/vk_descriptor_set.cpp:100]: I don't see where you're using device in here.

**Comment by JonCampbell407** [comment]
> This has two places where the immutable samplers could be written. This
> is not the final code. It is just something to start discussion. Let
> me know what you think.
> 
> Discussion for #128.

I think doing it in WriteDescriptorSets() counts on the app violating the spec.  I think doing it when the descriptor set is allocated is the best choice.  Instead of doing it in Reassign(), I'd prefer Reassign() be left as is and done in a function that is called after Reassign().

I think we should have a bit in DeviceFeatures to control this.  That can be set based on enableRelocatableShaders.

**Comment by s-perron** [comment]
> > This has two places where the immutable samplers could be written. This
> > is not the final code. It is just something to start discussion. Let
> > me know what you think.
> > Discussion for #128.
> 
> I think doing it in WriteDescriptorSets() counts on the app violating the spec. I think doing it when the descriptor set is allocated is the best choice. Instead of doing it in Reassign(), I'd prefer Reassign() be left as is and done in a function that is called after Reassign().
> 
> I think we should have a bit in DeviceFeatures to control this. That can be set based on enableRelocatableShaders.

Sounds good.  I'll do that.

**Comment by s-perron** [comment]
V2: I've updated the PR to reflect what @JonCampbell407 suggested.  Let me know if any other changes are needed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f653d8e787ff64933b739355d4e212c547884715
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev bd6be6fbda3fd5331b56073fdd66d9cab1381fa9</li>
<li>PAL: dev cd4f8b28b5d923e8d65a1ef1d3dee5137a89d6bd</li>
<li>SPVGEN: dev 8705e87a8b4aaed55a89f79ed423612ca310bfdd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 68e5420f0ba94be9cae6fbac848d746592d03f42</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/221024)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31062/55256 (56.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24194/55256 (43.8%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31062/55256 (56.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24194/55256 (43.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31436/55256 (56.9%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 23820/55256 (43.1%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34187/55256 (61.9%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 21069/55256 (38.1%)</li>
</ul></details>


</ul></details>

**Comment by s-perron** [comment]
V3: Rebased

**Comment by s-perron** [comment]
@JonCampbell407 This is ready to be reviewed whenever you have time.  Thanks.

File: icd/api/vk_descriptor_set.cpp [hunk 3]
```diff
@@ -89,6 +89,25 @@ void DescriptorSet<numPalDevices>::Reassign(
...
 }
 
+// =====================================================================================================================
+// Writes the immutable samplers in the layout to memory.
+template <uint32_t numPalDevices>
+void DescriptorSet<numPalDevices>::WriteImmutableSamplers()
+{
+    for (uint32_t deviceIdx = 0; deviceIdx < numPalDevices; deviceIdx++)
+    {
+        for (uint32_t bindingIndex = 0; bindingIndex < Layout()->Info().count; ++bindingIndex) {
```
Comment (JonCampbell407) [inline icd/api/vk_descriptor_set.cpp:99]: { on its own line.

File: icd/api/vk_descriptor_set.cpp [hunk 4]
```diff
@@ -89,6 +89,25 @@ void DescriptorSet<numPalDevices>::Reassign(
...
+          const DescriptorSetLayout::BindingInfo& bindingInfo = Layout()->Binding(bindingIndex);
+
+          if (bindingInfo.imm.dwSize == 0) continue;
```
Comment (JonCampbell407) [inline icd/api/vk_descriptor_set.cpp:102]: Instead of using continue test != 0.

File: icd/api/vk_descriptor_set.cpp [hunk 5]
```diff
@@ -89,6 +89,25 @@ void DescriptorSet<numPalDevices>::Reassign(
...
+
+          uint32_t * pSamplerDesc =(Layout()->Info().imm.pImmutableSamplerData + bindingInfo.imm.dwOffset);
```
Comment (JonCampbell407) [inline icd/api/vk_descriptor_set.cpp:104]: Extra space before *.  There is an extra set of () - some places we want the unnecessary () and others we don't.

File: icd/api/vk_descriptor_set.cpp (hunk 1 above)
Comment (JonCampbell407) [reply icd/api/vk_descriptor_set.cpp:106]: Use sizeof(uint) instead of 4.

**Comment by s-perron** [comment]
V4: Fixed formatting.

File: icd/api/vk_descriptor_set.cpp (hunk 3 above)
Comment (s-perron) [reply icd/api/vk_descriptor_set.cpp]: done.

File: icd/api/vk_descriptor_set.cpp (hunk 4 above)
Comment (s-perron) [reply icd/api/vk_descriptor_set.cpp]: Done.

File: icd/api/vk_descriptor_set.cpp (hunk 5 above)
Comment (s-perron) [reply icd/api/vk_descriptor_set.cpp]: Done.

**Review by JonCampbell407 — DISMISSED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7fd261a6d9d98d742a6f9003aaede5cc99e735d6
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev 82a8501c77b2e70488d11d213c583df0c56e617e</li>
<li>PAL: dev 8e5baae7cfdd885974013337e556e9dc955b6737</li>
<li>SPVGEN: dev 8705e87a8b4aaed55a89f79ed423612ca310bfdd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 68e5420f0ba94be9cae6fbac848d746592d03f42</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/221024)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31225/55256 (56.5%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24031/55256 (43.5%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31225/55256 (56.5%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24031/55256 (43.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31591/55256 (57.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 23665/55256 (42.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34350/55256 (62.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 20906/55256 (37.8%)</li>
</ul></details>


</ul></details>

File: icd/api/vk_descriptor_set.cpp [hunk 6]
```diff
@@ -89,6 +89,27 @@ void DescriptorSet<numPalDevices>::Reassign(
...
+{
+    for (uint32_t deviceIdx = 0; deviceIdx < numPalDevices; deviceIdx++)
+    {
+        for (uint32_t bindingIndex = 0; bindingIndex < Layout()->Info().count; ++bindingIndex)
+        {
+            const DescriptorSetLayout::BindingInfo& bindingInfo = Layout()->Binding(bindingIndex);
+
+            if (bindingInfo.imm.dwSize != 0)
+            {
+                uint32_t* pSamplerDesc =(Layout()->Info().imm.pImmutableSamplerData + bindingInfo.imm.dwOffset);
```
Comment (JaxLinAMD) [inline icd/api/vk_descriptor_set.cpp:105]: Hi @s-perron  please remove the extra parentheses

Comment (JaxLinAMD) [reply icd/api/vk_descriptor_set.cpp:105]: please add a space between "=" and "Layout"

Comment (s-perron) [reply icd/api/vk_descriptor_set.cpp]: Done.

**Comment by s-perron** [comment]
V5: More formatting fixes.

**Comment by amdvlk-admin** [comment]
### Test summary for commit fc202d9bcca540ab2b12468150f1def456b51aad
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev 78cfd426cc0eb45dcbfc53651b219a9172a4d456</li>
<li>PAL: dev 8e5baae7cfdd885974013337e556e9dc955b6737</li>
<li>SPVGEN: dev 8705e87a8b4aaed55a89f79ed423612ca310bfdd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 68e5420f0ba94be9cae6fbac848d746592d03f42</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/221024)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31225/55256 (56.5%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24031/55256 (43.5%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31225/55256 (56.5%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 24031/55256 (43.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31591/55256 (57.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 23665/55256 (42.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34350/55256 (62.2%)</li>
<li>Failed: 0/55256 (0.0%)</li>
<li>Not Supported: 20906/55256 (37.8%)</li>
</ul></details>


</ul></details>

File: icd/api/vk_descriptor_set.cpp (hunk 1 above)
Comment (JaxLinAMD) [reply icd/api/vk_descriptor_set.cpp]: @s-perron please change 4 to sizeof(uint) ( or maybe sizeof(uint32_t)? )

**Comment by s-perron** [comment]
V6: Use sizeof(uint) for the double word size.

File: icd/api/vk_descriptor_set.cpp [hunk 7]
```diff
@@ -89,6 +89,27 @@ void DescriptorSet<numPalDevices>::Reassign(
...
+    {
+        for (uint32_t bindingIndex = 0; bindingIndex < Layout()->Info().count; ++bindingIndex)
+        {
+            const DescriptorSetLayout::BindingInfo& bindingInfo = Layout()->Binding(bindingIndex);
+
+            if (bindingInfo.imm.dwSize != 0)
+            {
+                uint32_t* pSamplerDesc = Layout()->Info().imm.pImmutableSamplerData + bindingInfo.imm.dwOffset;
+                uint32_t* pDestAddr = StaticCpuAddress(deviceIdx) + Layout()->GetDstStaOffset(bindingInfo, 0);
+                memcpy(pDestAddr, pSamplerDesc, sizeof(uint) * bindingInfo.imm.dwSize);
```
Comment (JaxLinAMD) [inline icd/api/vk_descriptor_set.cpp:107]:
```suggestion
                memcpy(pDestAddr, pSamplerDesc, sizeof(uint32) * bindingInfo.imm.dwSize);
```

**Pull request merged**